### PR TITLE
Add ranges to `Violation`

### LINKED ISSUE #44: `Violation` should have `start` and `end`
Rather than:

```rust
pub enum ViolationPosition {
    None,
    Line(usize),
    LineCol((usize, usize)),
}
```

which just points to a specific place in the source, having a start and end allows pointing to exactly the offending piece of code and underlining the relevant part(s)

I'm not sure what `None` is supposed to represent here, but it could be encoded as `(0, 0)`. Similarly, `Line` could be the start & end of the whole line.

It would also be most useful if this was in terms of offset into the file, with some conversion utility between offset and line/column numbers. (This is at least partly what all that extra structure is for in ruff)

### PR DESCRIPTION
**Pull request by ZedThree**
Closes #44 

Uses the `ruff_source_file` and `ruff_text_size` crates to implement ranges for violations. Among other things, using these directly should make it easier to adapt other parts of ruff.

A bit rough but it works. Required some rejigging of the tests that I'm not overly happy with -- namely using zero-indexed line/column for the start/end ranges. Might be better to just use offsets in the tests as well, but then we probably want something to make it easier to work out the expected ranges. Maybe a way to optionally print the parse tree for the test snippet? Or to print the snippet with line-start-offsets, e.g.:

```
     0    5    1
               0   
0  | program test
13 |   implicit none
30 | end program test
```

### CODE REVIEW

File: src/check.rs
Comment (LiamPattinson) [inline src/check.rs:130]: I'd prefer to fix this now rather than later. Maybe something like this:

```suggestion
            let err_source = "kj23brdf9kshb23owihqb";
            for path in get_files(&args.files) {
                let source = read_to_string(&path).unwrap_or(err_source.to_string());
                if source == err_source {
                    let file = SourceFileBuilder::new(path.to_string_lossy().as_ref(), "")
                        .finish();
                    let err_msg = format!("Failed to process: {}", path.display());
                    let violation = violation!(&err_msg);
                    let diagnostic = Diagnostic::new(&file, "E000", &violation);
                    println!("{}", diagnostic);
                    total_errors += 1;
                    continue;
                }
```
The diagnostic handling bit is very similar to a section lower down in this function, so should probably be pulled out into a new function.

File: src/lib.rs
Comment (LiamPattinson) [inline src/lib.rs:91]: These unwraps feel a bit scary. Should this instead return `Result<Self>` and be handled with `?` everywhere?

File: src/lib.rs
Comment (LiamPattinson) [inline src/lib.rs:249]:
```suggestion
fn format_violation_range(
```

File: src/lib.rs
Comment (LiamPattinson) [inline src/lib.rs:301]: Can this be removed?

File: src/lib.rs
Comment (LiamPattinson) [inline src/lib.rs:299]: This refers back to earlier comments that were removed. Can this go too, or does it need to be reworded?

**Review by LiamPattinson — COMMENTED** [review]
All generally looks good, with just a few comments.

Overall, I'm a little uneasy with the direct use of Ruff internals from a maintainability point of view. It makes it harder for others to contribute if they have to figure out how to use largely-undocumented internals of another project (although it's not like ours is particularly well documented...).

On the other hand, this _works_, and I'd much prefer to make fast progress now than to get bogged down essentially reimplementing the same features ourselves. If it becomes a major headache in future, I imagine we'll be able to swap out the Ruff code for our own implementation fairly easily, especially if we aim for a similar API.

**Comment by LiamPattinson** [comment]
Should we also swap around where we write the error code and where we write the message in our diagnostics?

File: src/check.rs
Comment (ZedThree) [reply src/check.rs:130]: Yep, or pull the loop body out into a function that returns `Result`, and then can propagate the error up like that?

File: src/lib.rs
Comment (ZedThree) [reply src/lib.rs:91]: What we can do is check the file length as soon as we open it to ensure it's representable as `u32` (i.e. < 4GB), and then we know all these casts/unwraps are fine.

File: src/check.rs
Comment (LiamPattinson) [reply src/check.rs:130]: Yes, I think that would be a better approach

File: src/lib.rs
Comment (LiamPattinson) [reply src/lib.rs:91]: Ah, if that's the only time that this can panic, then it should be fine to leave them in, although maybe with a comment to explain why.

File: src/check.rs
Comment (ZedThree) [inline src/check.rs:148]: Can probably do this without calling `.len()` by asking the filesystem

**Review by LiamPattinson — APPROVED** [review]
All looks good to me! Cheers for getting this in, it should make things like auto-fixes much easier to implement. Up to you if you'd like to make a few final tweaks before merging.

**Comment by ZedThree** [comment]
Ok switched to using `Metadata.len()` so we don't need to actually open the file, and gave it a vaguely informative error message

**Comment by LiamPattinson** [comment]
Nice, looks like a really clean solution. Happy to merge if you are.

**Pull request merged**
